File: OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/gamifywidgetsdk/WidgetBottomSheet.java
package com.optimove.android.gamifywidgetsdk;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import org.json.JSONException;
import org.json.JSONObject;

class WidgetBottomSheet extends BottomSheetDialogFragment {

    private static final String TAG = "GamifyWidget";
    static final String FRAGMENT_TAG = "WidgetBottomSheet";
    private static final String ARG_WIDGET_URL = "widget_url";
    private static final String ARG_USER_ID = "user_id";
    private static final String ARG_TOKEN = "token";

    private WebView webView;
    private ProgressBar progressBar;
    private TextView errorView;

    static WidgetBottomSheet newInstance(String widgetUrl,
                                         @Nullable String userId,
                                         @Nullable String token) {
        WidgetBottomSheet sheet = new WidgetBottomSheet();
        Bundle args = new Bundle();
        args.putString(ARG_WIDGET_URL, widgetUrl);
        if (userId != null) {
            args.putString(ARG_USER_ID, userId);
        }
        if (token != null) {
            args.putString(ARG_TOKEN, token);
        }
        sheet.setArguments(args);
        return sheet;
    }

    private String getWidgetUrl() {
        Bundle args = getArguments();
        return args != null ? args.getString(ARG_WIDGET_URL, "") : "";
    }

    @Nullable
    private String getUserId() {
        Bundle args = getArguments();
        return args != null ? args.getString(ARG_USER_ID) : null;
    }

    @Nullable
    private String getToken() {
        Bundle args = getArguments();
        return args != null ? args.getString(ARG_TOKEN) : null;
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        FrameLayout root = new FrameLayout(requireContext());

        webView = new WebView(requireContext());
        webView.setLayoutParams(new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT));
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setLoadWithOverviewMode(true);
        settings.setUseWideViewPort(true);
        webView.addJavascriptInterface(
                new AndroidBridge(
                        new Runnable() {
                            @Override
                            public void run() {
                                Activity activity = getActivity();
                                if (activity != null) {
                                    activity.runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            dismiss();
                                        }
                                    });
                                }
                            }
                        },
                        new Runnable() {
                            @Override
                            public void run() {
                                Activity activity = getActivity();
                                if (activity != null) {
                                    activity.runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            sendInit();
                                        }
                                    });
                                }
                            }
                        }),
                "AndroidBridge");
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                Activity activity = getActivity();
                if (activity != null) {
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            progressBar.setVisibility(View.GONE);
                        }
                    });
                }
            }

            @Override
            public void onReceivedError(WebView view,
                                        WebResourceRequest request,
                                        WebResourceError error) {
                if (request.isForMainFrame()) {
                    Activity activity = getActivity();
                    if (activity != null) {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                showError();
                            }
                        });
                    }
                }
            }
        });

        progressBar = new ProgressBar(requireContext());
        progressBar.setLayoutParams(new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT,
                Gravity.CENTER));

        errorView = new TextView(requireContext());
        errorView.setLayoutParams(new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT,
                Gravity.CENTER));
        errorView.setText("Unable to load widget.\nCheck your connection and try again.");
        errorView.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
        errorView.setVisibility(View.GONE);

        root.addView(webView);
        root.addView(progressBar);
        root.addView(errorView);

        webView.loadUrl(getWidgetUrl());
        return root;
    }

    @Override
    public void onStart() {
        super.onStart();
        BottomSheetDialog dialog = (BottomSheetDialog) getDialog();
        if (dialog == null) {
            return;
        }
        FrameLayout bottomSheet = dialog.findViewById(
                com.google.android.material.R.id.design_bottom_sheet);
        if (bottomSheet == null) {
            return;
        }
        BottomSheetBehavior<FrameLayout> behavior = BottomSheetBehavior.from(bottomSheet);
        behavior.setState(BottomSheetBehavior.STATE_EXPANDED);
        behavior.setSkipCollapsed(true);
    }

    /**
     * Called by AndroidBridge.onReady() when the widget fires its READY signal.
     * Posts the INIT payload back via window.postMessage so the widget bridge picks it up.
     */
    private void sendInit() {
        String initJson = buildInitJson();
        Log.d(TAG, "sending INIT: " + redactedLog(initJson));
        webView.evaluateJavascript("window.postMessage(" + initJson + ", '*');", null);
    }

    private String redactedLog(String json) {
        try {
            JSONObject obj = new JSONObject(json);
            if (obj.has("token")) {
                obj.put("token", "[REDACTED]");
            }
            return obj.toString();
        } catch (JSONException e) {
            return "[unparseable]";
        }
    }

    private String buildInitJson() {
        try {
            JSONObject obj = new JSONObject();
            obj.put("type", "INIT");
            String userId = getUserId();
            if (userId != null) {
                obj.put("userId", userId);
            }
            String token = getToken();
            if (token != null) {
                obj.put("token", token);
            }
            return obj.toString();
        } catch (JSONException e) {
            return "{\"type\":\"INIT\"}";
        }
    }

    private void showError() {
        progressBar.setVisibility(View.GONE);
        webView.setVisibility(View.GONE);
        errorView.setVisibility(View.VISIBLE);
    }
}

File: OptimoveSDK/gamify-widget-sdk/src/main/java/com/optimove/android/gamifywidgetsdk/WidgetBottomSheet.kt
package com.optimove.android.gamifywidgetsdk

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import android.util.Log
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import org.json.JSONObject

internal class WidgetBottomSheet : BottomSheetDialogFragment() {

    private lateinit var webView: WebView
    private lateinit var progressBar: ProgressBar
    private lateinit var errorView: TextView

    private val widgetUrl: String get() = arguments?.getString(ARG_WIDGET_URL) ?: ""
    private val userId: String? get() = arguments?.getString(ARG_USER_ID)
    private val token: String? get() = arguments?.getString(ARG_TOKEN)

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val root = FrameLayout(requireContext())

        // WebView
        webView = WebView(requireContext()).apply {
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
            settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true
                loadWithOverviewMode = true
                useWideViewPort = true
            }
            addJavascriptInterface(
                AndroidBridge(
                    onClose = { activity?.runOnUiThread { dismiss() } },
                    onReady = { activity?.runOnUiThread { sendInit() } }
                ),
                "AndroidBridge"
            )
            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView, url: String) {
                    activity?.runOnUiThread { progressBar.visibility = View.GONE }
                }

                override fun onReceivedError(
                    view: WebView,
                    request: WebResourceRequest,
                    error: WebResourceError
                ) {
                    if (request.isForMainFrame) {
                        activity?.runOnUiThread { showError() }
                    }
                }
            }
        }

        // Loading spinner
        progressBar = ProgressBar(requireContext()).apply {
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT,
                android.view.Gravity.CENTER
            )
        }

        // Error state
        errorView = TextView(requireContext()).apply {
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT,
                android.view.Gravity.CENTER
            )
            text = "Unable to load widget.\nCheck your connection and try again."
            textAlignment = View.TEXT_ALIGNMENT_CENTER
            visibility = View.GONE
        }

        root.addView(webView)
        root.addView(progressBar)
        root.addView(errorView)

        webView.loadUrl(widgetUrl)
        return root
    }

    override fun onStart() {
        super.onStart()
        // Expand the bottom sheet fully on open
        val bottomSheet = (dialog as? BottomSheetDialog)
            ?.findViewById<FrameLayout>(com.google.android.material.R.id.design_bottom_sheet)
        bottomSheet?.let {
            BottomSheetBehavior.from(it).apply {
                state = BottomSheetBehavior.STATE_EXPANDED
                skipCollapsed = true
            }
        }
    }

    /**
     * Called by AndroidBridge.onReady() when the widget fires its READY signal.
     * Posts the INIT payload back via window.postMessage so useBridge picks it up.
     */
    private fun sendInit() {
        val initJson = buildInitJson()
        Log.d(TAG, "sending INIT: $initJson")
        webView.evaluateJavascript("window.postMessage($initJson, '*');", null)
    }

    private fun buildInitJson(): String {
        val obj = JSONObject().apply {
            put("type", "INIT")
            userId?.let { put("userId", it) }
            token?.let { put("token", it) }
        }
        return obj.toString()
    }

    private fun showError() {
        progressBar.visibility = View.GONE
        webView.visibility = View.GONE
        errorView.visibility = View.VISIBLE
    }

    companion object {
        private const val TAG = "GamifyWidget"
        const val FRAGMENT_TAG = "WidgetBottomSheet"
        private const val ARG_WIDGET_URL = "widget_url"
        private const val ARG_USER_ID = "user_id"
        private const val ARG_TOKEN = "token"

        fun newInstance(widgetUrl: String, userId: String?, token: String?): WidgetBottomSheet {
            return WidgetBottomSheet().apply {
                arguments = Bundle().apply {
                    putString(ARG_WIDGET_URL, widgetUrl)
                    userId?.let { putString(ARG_USER_ID, it) }
                    token?.let { putString(ARG_TOKEN, it) }
                }
            }
        }
    }
}
